keel-core/src/main/kotlin/com/netflix/spinnaker/keel/jackson/SerializationExtensionRegistry.kt

```diff
@@ -0,0 +1,40 @@
+package com.netflix.spinnaker.keel.jackson
+
+import com.fasterxml.jackson.databind.JsonDeserializer
+import com.fasterxml.jackson.databind.JsonSerializer
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+/**
+ * A registry to allow customization of JSON serializers and deserializers.
+ */
+@Component
+class SerializationExtensionRegistry {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  private val _deserializers: MutableMap<Class<*>, JsonDeserializer<*>> = mutableMapOf()
+  private val _serializers: MutableMap<Class<*>, JsonSerializer<*>> = mutableMapOf()
+
+  val deserializers: Map<Class<*>, JsonDeserializer<*>>
+    get() = _deserializers.toMap()
+
+  val serializers: Map<Class<*>, JsonSerializer<*>>
+    get() = _serializers.toMap()
+
+  fun <T : JsonDeserializer<*>> register(javaClass: Class<*>, deserializer: T) {
+    _deserializers.putIfAbsent(javaClass, deserializer)
+      ?.let {
+        if (it.javaClass != deserializer.javaClass) {
+          log.warn("Attempt to register deserializer ${deserializer.javaClass.name} for type ${javaClass.name}, but ${it.javaClass.name} is already registered.")
```

Comment:
Doesn't that prevent the method from being idempotent if you call it with the same parameters though?

Reply:
No, because the warning is only printed if the class of the parameter is different (line 26). If you pass a different type, then it's ignored because I wanted the first registration to take precedence, hence the warning in that case.

Reply:
ok, I see

Reply:
ok, I see